### ISSUE TITLE
NaNs in SSL

### ISSUE THREAD

**Issue by sophie-xhonneux** [ecmwf/WeatherGenerator#2721]
# Description

When running space JEPA, we get NaNs. For example:
```
/e/home/jusers/xhonneux2/jupiter/projects/dev-ssl/WeatherGenerator/src/weathergen/train/trainer.py:1304: RuntimeWarning: Mean of empty slice
  + f"{self.cf.general.istep:06d} : loss = {np.nanmean(avg_loss):.4E} "
000 : 00010/02046 : 000010 : loss = NAN (lr=1.23E-06, s/sec=0.058)
	
LossPhysical.ERA5.mse.avg : 8.4205E-01 	
LossPhysical.loss_avg : 8.4205E-01 	
LossLatentSSLStudentTeacher.loss_avg : nan 	

000 : 00020/02046 : 000020 : loss = NAN (lr=1.83E-06, s/sec=0.088)
	
LossPhysical.ERA5.mse.avg : 7.9669E-01 	
LossPhysical.loss_avg : 7.9669E-01 	
LossLatentSSLStudentTeacher.loss_avg : nan
```

Using Fable to analyse it was found that they come from uninitialized GPU memory: EmbeddingEngine.forward (src/weathergen/model/engines.py:89) allocates tokens_all = torch.empty(num_tokens, dim) and fills it with scatter_, but num_tokens counts tokens from all streams while the embed loop skips streams with an nn.Identity embed — i.e. your target-only, diagnostic ERA5 stream (diagnostic: true / sources_size == 0, engines.py:55-57). In the teacher's target view, that ERA5 stream contributes ~52,000 token rows that are never written and stay as garbage memory. Random bit patterns make ~0.2% of those elements NaN; one NaN is enough — local assimilation spreads it to ~65% of cells and the aggregation engine's global attention to 100% of teacher tokens. The student's model_input view has no ERA5 source tokens (its rows are fully covered), so LossPhysical stays finite while LossLatentSSLStudentTeacher goes NaN. I reproduced and instrumented this end-to-end on a free GH200 on this node — on every teacher forward: num_tokens=108632, nonfinite=113187, bad_rows=52174, skipped=[('ERA5','identity')], while every student forward is clean.

Why the symptoms looked the way they did

- Intermittent ~57% on the cluster: whether the garbage memory happens to contain NaN bit patterns depends on allocator reuse — nondeterministic per batch. On my repro node it was 100%.
- Worse than NaN: on the "clean" 43% of batches the skipped rows are finite garbage, so the teacher's JEPA/context targets have been silently corrupted on every batch of these runs. The final ejtgnnxu checkpoint was also written after ~450 skipped-batch steps of this.
- "Starts at batch 33" is an artifact: spike detection needs min_history: 20 finite losses before it can flag anything; the NaNs are there from batch 0 (before that, only the GradScaler silently rejected them).
- Cross-run evidence fits perfectly: every NaN run (ejtgnnxu, op7gx2tw, tcyznkb1, g3xl7ae1, makg7qlc, tlial69l, txwyjh18, de3z9l3a, jc7v6b5d) uses a stream set containing the source-less diagnostic ERA5 stream; every clean deep-SSL run (ci8dbx75, cunkz657, mn401qfs, …) doesn't.
- EMA ramp-up is exonerated: tcyznkb1/makg7qlc/g3xl7ae1 ran with ema_ramp_up_ratio: null and still NaN'd from batch 10, and my repro NaNs on the very first batch before any EMA update. I also loaded ejtgnnxu_latest.chkpt + .ema_teacher — all 2,823 tensors finite, so no parameter poisoning. If your ramp-null run looks stable, check results/<run_id>/loss_spikes.jsonl and the per-window LossLatentSSLStudentTeacher.loss_avg lines — the terminal loss = value is a nanmean and hides NaN batches.

Recommended fix

1. Real fix: don't feed Identity-embed (diagnostic / source: []) streams' tokens into the encoder input at all — the teacher's target-view samples shouldn't tokenize ERA5-out as encoder input (this is the same territory as your 5a41e35a source/target-list fix).
2. Safety net either way in EmbeddingEngine.forward: use torch.zeros instead of torch.empty, and/or assert that the scattered row count equals num_tokens so incomplete coverage can never be silent again.
